rewrite to rescript; change doc comments to @ocaml.doc

LaMavia opened the pull request.

Commit:
rewrite to rescript; change doc comments to @ocaml.doc

ncthbrt commented on src/Express.res:
```diff
@@ -0,0 +1,624 @@
+  type t = (Js.undefined<content>, Response.t) => complete
+  let middleware: Js.undefined<content>
+
+  @ocaml.doc("value to use as [next] callback argument to invoke the next middleware")
```
Might have been like this before, but the doc comments don't seem to line up with the functionality. For the implementation `*.res` file, actually don't think the comments are necessary anyway

LaMavia replied:
Ok, I see. They might've shifted during the reformatting or it might've been just a silly mistake. I agree that the comments are not necessary in the implementation, so how about I remove them and make sure, this time, that they do match in the interface?

ncthbrt replied:
Perfect. Thanks a lot!

ncthbrt commented:
I'm going to cut a release once we've got this change in. We should probably bump the major package version up as the upgrade to the lates bs-platform is probably a breaking change

Commit:
remove comments from the implementation; remove odd spaces from the comments; move Static.t to the top of the module

DerivedMate commented:
While we're on the topic of breaking changes: has there been any discussion on abandoning `@bs.send.pipe` in favour of `@bs.send`? The former appears to be deprecated, together with the `|>` operator

ncthbrt commented:
I'd be in favour of it. There is no reason not to keep up with current standards

Commit:
rewrite to pipe-first; add comments for functions not covered in the tests

DerivedMate commented:
I've yet to rewrite non-external functions to pipe-first, but I doubt I'll have time to do it in the next week or two

ncthbrt commented:
@DerivedMate Sorry to clarify, would you say that this is ready for merge?

DerivedMate commented:
I thought about rewriting `next` to use `res` in a pipe-first manner, but now that I think about, `res` in this case is used more as a parameter than the main object (as opposed to the way it's used in `Response` methods). So, yes, I'd say it's ready to be merged; unless you see anything worth changing

ncthbrt commented:
If you're happy, I'll review more thoroughly and then try to cut a release. Thank you!

Commit:
remove needless parentheses

Commit:
rewrite Request.[accepts, acceptsCharsets], Response.[cookie, clearCookie] to pipe-first

DerivedMate commented:
Sorry to bother you, but I've encountered another thing that may need changing: I was trying to help with #89 but I encountered a few problems along the way:
1. "Cannot find module 'pug' [...]". First of all, it was difficult to diagnose as Ocaml errors are not logged as objects but as `[object object]`. Unfortunately, I wasn't able to fix the logging, as it appears to be a part of express itself: application.js
    ```javascript
    function logerror(err) {
      /* istanbul ignore next */
      if (this.get('env') !== 'test') console.error(err.stack || err.toString());
    }
    ```   
     The only fix I was able to find was adding `app.engine('pug', require('pug').__express)`which was not yet bound, so I'd suggest the following binding, should a similar problem arise for other engines:
     ```reason
       @bs.send external engine : (t, string, 'engine) => unit = "engine"
     ```
    This would allow for such code, which, in my opinion, is as good as it's gonna get unless dedicated engine bindings are created: 
   ```ocaml
      open Express
      let pug = [%raw {|require('pug').__express|}]
      
      let () = 
         let app = express ()
         in app |. App.engine "pug" pug           ;
            app |. App.set "view engine" "pug"    ;
            app |. App.set "views" "./views"      ;   
           (* ... *)
   ```

2. Passing parameters to `Response.render`. As mentioned in the issue (which, apparently, was closed while I was still writing this comment...), the parameters need to be of type `Js.Dict.t<string>`, which is limiting, not to say annoying. Any requests sent by the server (such as database queries) tend to be serialized into records; therefore, I'd suggest changing the binding from dict to `'v`.

Combining the changes from the two issues, it'd allow for:
```ocaml
open Express 

type index_params = 
  { title : string
  ; names : string array 
  }

let pug = [%raw {|require('pug').__express|}]

let () = 
  let app = express()
  in app |. App.engine "pug" pug        ;
     app |. App.set "view engine" "pug" ;
     app |. App.set "views" "./views"   ;
    
     App.get app ~path:"/" 
         @@ Middleware.from (fun _next _req res -> 
          let dict: index_params = 
            { title = "Home"
            ; names = [|"Bob"; "Arthur"|]
            }
          in Response.render res "index" dict ()
         ) ;
   
   let _server = app |. App.listen 
                        ~port:8080 
                        ~onListen:(fun x -> Js.log x)
                        ()
   in ()
```

ncthbrt commented:
That makes sense. These bindings were created at a time when records did not correspond to how you'd expect them to be represented in JSON (think they were arrays)

Commit:
+App.engine, Response.render: Js.Dict.t -> 'v, bump bs-platform@8.3.1

dusty-phillips commented:
any thoughts on what it will take to ship this? I'm willing to take it over if changes are required.

ncthbrt commented:
@dusty-phillips: I'm unsure. I was waiting on @DerivedMate's feedback.

DerivedMate commented:
I thought it was ready to be merged, and I was waiting for feedback

ncthbrt commented:
Sorry @DerivedMate 😂. I didn't see that last commit on this PR

ncthbrt commented:
@dusty-phillips: Are you able to test this PR? As I mentioned in #80, I'm not actively keeping up to date with the ReScript/Reason community.

dusty-phillips commented:
I'll give it a shot. Might take a bit, as I'm brand new to Rescript and not sure how to run against a package locally, etc, but I'll figure it out!

dusty-phillips commented:
@DerivedMate unit tests are failing for me, but there isn't any output to indicate why. Do they pass for you, or is this a local issue?

DerivedMate commented:
@dusty-phillips yes, I get the same thing locally but it all passes on travis.
Give me a moment, I wonder if the same happens with the current master

DerivedMate commented:
The same appears to apply for the master:
```
> status: 404
181c181
< status: 200# Netscape HTTP Cookie File
---
> status: 200
194c194
< status: 404# Netscape HTTP Cookie File
---
> status: 404
198c198
< status: 200# Netscape HTTP Cookie File
---
> status: 200
202c202
< status: 201# Netscape HTTP Cookie File
---
> status: 201
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! bs-express@1.0.2 test: `cd tests && ./test.sh`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the bs-express@1.0.2 test script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

```

dusty-phillips committed:
Migrate the example code to Rescript

@DerivedMate has ported bs-express to Rescript,
but the example is still in Reason.

I've ported it to Rescript syntax by using the
rescript playground to convert from reason to Rescript
and then hand editing the parts that "looked weird".

I have removed the original Reason file, but we could also
leave it in if you think there is benefit to having
two examples.

dusty-phillips commented:
@ncthbrt I've played around with this PR all afternoon and haven't been able to find any issues with it. I think it's safe to publish as a breaking change. I'm brand new to Rescript, though, so it may be worth taking that with a grain of salt.

I updated the examples file from Reason to Rescript in to see how it would look. I'm not sure if you want to include that file, but I submitted it as a PR to @DerivedMate's original branch here: https://github.com/DerivedMate/bs-express/pull/1

LaMavia committed:
Setted -> set 

A small grammar correction. It must have been there for a while.

LaMavia committed:
Merge pull request #1 from dusty-phillips/res

Migrate the example code to Rescript

Commit:
bump: body-parser, bs-json, bs-platform, express

dusty-phillips commented:
@ncthbrt can we get a merge? :-)

ncthbrt commented:
We can

ncthbrt commented:
Will still need to manually cut a release. Travis is broken I believe

dusty-phillips commented:
What does that involve? I'm not intimate with npm; is it something one of us can help out with (I'm volunteering @DerivedMate because I don't know where to start :-D)?

ncthbrt commented:
I've annoyingly had to do that on a number of the repos I'm the maintainer of. The best long term fix entails writing a Github Action that replicates the original pipeline. I can cut a release in the meantime though. I can do it this evening.
